File: bit_login/sso/fingerprint.py
import hashlib
import json
from dataclasses import dataclass
from typing import Any, Sequence


def _js_json(value: Any) -> str:
    if value is None:
        return "undefined"
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def _sha256(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


@dataclass(frozen=True)
class BrowserFingerprintProfile:
    """Non-secret browser characteristics used to create a fresh risk profile."""

    fonts: Sequence[str] = (
        "Arial",
        "Helvetica Neue",
        "PingFang SC",
        "Times New Roman",
    )
    device_memory: int = 16
    hardware_concurrency: int = 10
    timezone: str = "Asia/Shanghai"
    cpu_class: str = "not available"
    platform: str = "MacIntel"
    language: str = "zh-CN"
    screen_resolution: tuple[int, int] = (956, 1470)
    platform_authenticator: str = "support"

    def build(self, *, cookie_value: str, user_agent: str, group_id: str = "") -> dict[str, str]:
        values = {
            "fonts": _js_json(list(self.fonts)),
            "deviceMemory": _js_json(self.device_memory),
            "hardwareConcurrency": _js_json(self.hardware_concurrency),
            "timezone": _js_json(self.timezone),
            "cpuClass": _js_json(self.cpu_class),
            "platform": _js_json(self.platform),
            "language": _js_json(self.language),
            "screenResolution": _js_json(list(self.screen_resolution)),
        }
        combined = "".join(values.values())
        return {
            "fonts": _sha256(values["fonts"]),
            "deviceMemory": _sha256(values["deviceMemory"]),
            "hardwareConcurrency": _sha256(values["hardwareConcurrency"]),
            "localgroupId": group_id,
            "timezone": values["timezone"],
            "cpuClass": _sha256(values["cpuClass"]),
            "platform": values["platform"],
            "language": values["language"],
            "screenResolution": values["screenResolution"],
            "fingerprint": _sha256(combined),
            "cookieValue": cookie_value,
            "userAgent": user_agent,
            "platformAuthenticator": self.platform_authenticator,
        }

File: bit-login/src/commonMain/kotlin/cn/bit101/bitlogin/sso/Fingerprint.kt
package cn.bit101.bitlogin.sso

data class BrowserFingerprintProfile(
    val fonts: List<String> = DEFAULT_FONTS,
    val deviceMemory: Int = 16,
    val hardwareConcurrency: Int = 10,
    val timezone: String = "Asia/Shanghai",
    val cpuClass: String = "not available",
    val platform: String = "MacIntel",
    val language: String = "zh-CN",
    val screenResolution: List<Int> = DEFAULT_SCREEN_RESOLUTION,
    val platformAuthenticator: String = "support",
) {
    fun build(cookieValue: String, userAgent: String, groupId: String = ""): Map<String, String> {
        val values = mapOf(
            "fonts" to jsList(fonts),
            "deviceMemory" to deviceMemory.toString(),
            "hardwareConcurrency" to hardwareConcurrency.toString(),
            "timezone" to jsString(timezone),
            "cpuClass" to jsString(cpuClass),
            "platform" to jsString(platform),
            "language" to jsString(language),
            "screenResolution" to jsIntList(screenResolution),
        )
        val combined = values.values.joinToString("")
        return mapOf(
            "fonts" to sha256(values.getValue("fonts")),
            "deviceMemory" to sha256(values.getValue("deviceMemory")),
            "hardwareConcurrency" to sha256(values.getValue("hardwareConcurrency")),
            "localgroupId" to groupId,
            "timezone" to values.getValue("timezone"),
            "cpuClass" to sha256(values.getValue("cpuClass")),
            "platform" to values.getValue("platform"),
            "language" to values.getValue("language"),
            "screenResolution" to values.getValue("screenResolution"),
            "fingerprint" to sha256(combined),
            "cookieValue" to cookieValue,
            "userAgent" to userAgent,
            "platformAuthenticator" to platformAuthenticator,
        )
    }

    companion object {
        val DEFAULT_FONTS = listOf("Arial", "Helvetica Neue", "PingFang SC", "Times New Roman")
        val DEFAULT_SCREEN_RESOLUTION = listOf(956, 1470)

        private fun jsString(s: String): String = "\"" +
            s.replace("\\", "\\\\").replace("\"", "\\\"") + "\""

        private fun jsList(items: List<String>): String =
            "[" + items.joinToString(",") { jsString(it) } + "]"

        private fun jsIntList(items: List<Int>): String =
            "[" + items.joinToString(",") { it.toString() } + "]"

        private fun sha256(value: String): String =
            Crypto.sha256(value.encodeToByteArray())
    }
}
